mealie/core/security/security.py

```diff
@@ -47,23 +47,38 @@ def user_from_ldap(db: AllRepositories, session, username: str, password: str) -
 
     settings = get_app_settings()
 
+    if settings.LDAP_TLS_INSECURE:
+        ldap.set_option(ldap.OPT_X_TLS_REQUIRE_CERT, ldap.OPT_X_TLS_NEVER)
+
+    ldap.set_option(ldap.OPT_REFERRALS, 0)
+    ldap.set_option(ldap.OPT_PROTOCOL_VERSION, 3)
     conn = ldap.initialize(settings.LDAP_SERVER_URL)
-    user_dn = settings.LDAP_BIND_TEMPLATE.format(username)
+
+    if settings.LDAP_TLS_CACERTFILE:
+        conn.set_option(ldap.OPT_X_TLS_CACERTFILE, settings.LDAP_TLS_CACERTFILE)
+        conn.set_option(ldap.OPT_X_TLS_NEWCTX, 0)
+
+    user = db.users.get_one(username, "email", any_case=True)
+    if not user:
+        user_bind = settings.LDAP_BIND_TEMPLATE.format(username)
+        user = db.users.get_one(username, "username", any_case=True)
+    else:
+        user_bind = settings.LDAP_BIND_TEMPLATE.format(user.username)
+
     try:
-        conn.simple_bind_s(user_dn, password)
+        conn.simple_bind_s(user_bind, password)
     except (ldap.INVALID_CREDENTIALS, ldap.NO_SUCH_OBJECT):
         return False
 
-    user = db.users.get_one(username, "username", any_case=True)
+    user_dn, user_attr = conn.search_s(settings.LDAP_BASE_DN, ldap.SCOPE_SUBTREE, '(&(objectClass=user)(|(cn=%s)(sAMAccountName=%s)(mail=%s)))' % (username, username, username), ['name', 'mail'])[0]
```

Comment:
Can you provide a code comment to explain what/why this string is needed?

Reply:
Done and done.